askama_shared/src/generator.rs

```diff
@@ -719,21 +719,58 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
         let mut names = Buffer::new(0);
         let mut values = Buffer::new(0);
         for (i, arg) in def.args.iter().enumerate() {
-            if i > 0 {
-                names.write(", ");
-                values.write(", ");
+            let expr = args.get(i).ok_or_else(|| {
+                CompileError::String(format!("macro '{}' takes more than {} arguments", name, i))
+            })?;
+
+            match expr {
+                // If `expr` is already a form of variable then
+                // don't reintroduce a new variable. This is
+                // to avoid moving non-copyable values.
+                Expr::Var(name) => {
+                    let var = self
+                        .locals
+                        .resolve_var(name)
+                        .map(ToString::to_string)
+                        .unwrap_or_else(|| format!("self.{}", name));
+                    self.locals.insert(arg, Some(var));
+                }
+                Expr::Attr(obj, attr) => {
+                    let mut attr_buf = Buffer::new(0);
+                    self.visit_attr(&mut attr_buf, obj, attr)?;
+
+                    let var = self
+                        .locals
+                        .resolve_var(&attr_buf.buf)
+                        .map(ToString::to_string)
+                        .unwrap_or(attr_buf.buf);
+                    self.locals.insert(arg, Some(var));
+                    continue;
```

Comment:
The `continue` here isn't needed, I think?

Reply:
Hmm, weird. Must be left over from when I originally only had 1 arm, and didn't use a catchall pattern. Removed.